### PR TITLE
Update Z-score plot

### PR DESCRIPTION
**Pull request by azywot**
I updated `Z-score` plot, change it to the area plot and added the functionality we talked about (i.e. showing only the compounds as points that are outside pre-defined z-score range).

On `Z-score` range being changed the lines move  and when `Apply` button is clicked the changes are applied (points are shown).

Below I'm pasting the preview, highly encourage you to play with it/give some suggestions 🙏 (one thing we may decide upon is whether to sort the points according to `mean` or `std` value - as described on the screenshots it results in the main line/shaded area being respectively smoother/more spiky - **for now it is sorted wrt `std`**).

obraz
obraz
obraz
obraz

**NOTE 0:** I haven't done anything related to `act`/`inh` plots, sorry for the misleading branch name.
**NOTE 1:** z-score range on `RangeSlider` is computed basing on the values of compounds - in the beginning the range is maximal (includes all points, therefore no points are visible) + the button `Aply` is dsiabled until the range is computed.
**NOTE 2:** Due to the std of `COMPOUND POS` being very large, the initial view might be not pleasant (as visible on ss) but it can be switched off using the legend so imo it's not a problem.
**NOTE 3:** Last but not least - it takes a while 😉 (for the future we certainly need to add ↺, ⏳, 🔄 and stuff).

### CODE REVIEW

File: dashboard/visualization/z_score_plot.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
+import numpy as np
+import pandas as pd
+import plotly.graph_objs as go
+
+PLOTLY_TEMPLATE = "plotly_white"
+
+
+def prepare_data_plot(df, well, col, id_pos="id_pos", range_max=None):
```
Comment (Tremirre) [inline dashboard/visualization/z_score_plot.py:8]: I'd suggest renaming this function to indicate that it prepares the dataframe and not the actual plot

File: dashboard/visualization/z_score_plot.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+            x_positions.append(i)
+            i += frac
+        df[id_pos] = x_positions
+    else:
+        df[id_pos] = range(len(df))
+
+    return df
+
+
+def plot_zscore(
```
Comment (Tremirre) [inline dashboard/visualization/z_score_plot.py:33]: its' quite unintuitive for a plotting function to return a dataframe, moreover the function itself is quite complex
I'd highly recommend breaking it up into at least two parts:
- preparing dataframes for visualization
- making the actual visualization

thus making it less confusing

File: dashboard/pages/primary_screening/callbacks.py [hunk 3]
```diff
@@ -273,8 +276,28 @@ def on_summary_entry(current_stage: int, stored_uuid: str, file_storage: FileSto
...
+    min_z, max_z = compounds_df["Z-SCORE"].min(), compounds_df["Z-SCORE"].max()
+    marks = {i: "{}".format(i) for i in range(0, ceil(max_z), 5)}
+    if floor(min_z) < 0:
+        marks.update({i: "{}".format(i) for i in range(floor(min_z), 0, 5)})
+
+    # NOTE: here I only added the position that is required for the plot_zscore function
+    fig_z_score, new_echo_df = plot_zscore(compounds_df, control_pos_df, control_neg_df)
+
+    serialized_new_echo_df = new_echo_df.reset_index().to_parquet()
+    file_storage.save_file(f"{stored_uuid}_echo_df.pq", serialized_new_echo_df)
```
Comment (Tremirre) [inline dashboard/pages/primary_screening/callbacks.py:288]: What is the reason for saving changed echo_df? Won't it work without saving?

File: dashboard/pages/primary_screening/callbacks.py [hunk 4]
```diff
@@ -273,8 +276,28 @@ def on_summary_entry(current_stage: int, stored_uuid: str, file_storage: FileSto
 
     echo_bmg_combined = echo_bmg_combined.reset_index().to_dict("records")
```
Comment (Tremirre) [inline dashboard/pages/primary_screening/callbacks.py:277]: As of current this line crashes the server when run twice (on changed echo_df - e.g. when user goes back from `report` to `summary` or does not reupload the echo files)

File: dashboard/pages/primary_screening/callbacks.py [hunk 5]
```diff
@@ -285,62 +308,104 @@ def on_summary_entry(current_stage: int, stored_uuid: str, file_storage: FileSto
...
+    PLATE = "Destination Plate Barcode"
+    WELL = "Destination Well"
+    Z_SCORE = "Z-SCORE"
+
+    echo_df = pd.read_parquet(
+        pa.BufferReader(file_storage.read_file(f"{stored_uuid}_echo_df.pq"))
+    )
+
+    # TBD: use pa.ParquetDataset
+    compounds_outside_df = echo_df[
```
Comment (Tremirre) [inline dashboard/pages/primary_screening/callbacks.py:394]: For Readability I'd recommend extracting the mask to a variable,
e.g.
```python
compounds_outside_df = echo_df[not_null_mask & z_score_in_range_mask][...]
```

File: dashboard/visualization/z_score_plot.py [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+            )
+        )
+
+    fig.add_hline(
+        y=z_score_limits[0],
+        line_width=3,
+        line_dash="dash",
+        line_color="gray",
+        annotation_text=f"MIN: {z_score_limits[0]:.2f}",
+        annotation_font_color="gray",
```
Comment (Tremirre) [inline dashboard/visualization/z_score_plot.py:137]: the current styling makes it difficult to see the annotation when a lot of points are on the plot:

image.png

**Review by Tremirre — COMMENTED** [review]
The new plot is much better looking and works well
there still are performance issues, we may have to discuss with @dabrze the possible options
(Although we'd need to confirm that the bottleneck is on the plotly side and not in extensive dataframe manipulations)

I added a few styling/code suggestions
so far we were able to maintain bidirectional process traversal and it'd be nice to keep it that way - current implementation breaks that feature

\+ missing annotations

**Comment by dabrze** [comment]
Commenting from my cellphone (this might reduce the value if my comment). I wouldn't sort the x-axis according to values, I would sort according to plates and then rows or columns. Thus way one can spot a suspicious plate/row ir column. The area plot should span between min and max of the "ribbonned" values. We don't want it to be narrowwer than what it summarizes. By default I would set the area'd Z to +/- 3 or some other the value the experts like.

**Comment by dabrze** [comment]
Also, later on, make sure the hover text is interesting. Seeing valueable info on hover is always great - well info, compound id and EOS link, cgemical diagram, or whatever the experts would want to see.

**Review by dabrze — COMMENTED** [review]
Left a few comments in the discussion thread.

**Comment by azywot** [comment]
Changed the pipeline so now this one is irrelevant

**Pull request closed without merging**